cluster-autoscaler/core/scaledown/eligibility/eligibility.go

```diff
@@ -0,0 +1,182 @@
+	GetScaleDownUtilizationThreshold(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (float64, error)
+	// GetScaleDownGpuUtilizationThreshold returns ScaleDownGpuUtilizationThreshold value that should be used for a given NodeGroup.
+	GetScaleDownGpuUtilizationThreshold(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (float64, error)
+}
+
+// NewChecker creates a new Checker object.
+func NewChecker(context *context.AutoscalingContext, thresholdGetter utilizationThresholdProcessor, unremovableNodes *unremovable.Nodes) *Checker {
+	return &Checker{
+		context:          context,
+		thresholdGetter:  thresholdGetter,
+		unremovableNodes: unremovableNodes,
+	}
+}
+
+// FilterOutUnremovable accepts a list of nodes that are candidates for
+// scale down and filters out nodes that cannot be removed, along with node
+// utilization info.
+// TODO(x13n): Node utilization could actually be calculated independently for
+// all nodes and just used here. Next refactor...
+func (c *Checker) FilterOutUnremovable(scaleDownCandidates []*apiv1.Node, timestamp time.Time) ([]string, map[string]utilization.Info) {
+	c.unremovableNodes.Update(c.context.ClusterSnapshot.NodeInfos(), timestamp)
+
+	skipped := 0
+	utilizationMap := make(map[string]utilization.Info)
+	currentlyUnneededNodeNames := make([]string, 0, len(scaleDownCandidates))
+
+	for _, node := range scaleDownCandidates {
+		nodeInfo, err := c.context.ClusterSnapshot.NodeInfos().Get(node.Name)
+		if err != nil {
+			klog.Errorf("Can't retrieve scale-down candidate %s from snapshot, err: %v", node.Name, err)
+			c.unremovableNodes.AddReason(node, simulator.UnexpectedError)
+			continue
+		}
+
+		// Skip nodes that were recently checked.
+		if c.unremovableNodes.IsRecent(node.Name) {
+			c.unremovableNodes.AddReason(node, simulator.RecentlyUnremovable)
+			skipped++
+			continue
+		}
+
+		reason, utilInfo := c.checkNodeUtilization(timestamp, nodeInfo)
+		if utilInfo != nil {
+			utilizationMap[node.Name] = *utilInfo
+		}
+		if reason != simulator.NoReason {
+			c.unremovableNodes.AddReason(node, reason)
+			continue
+		}
+
+		currentlyUnneededNodeNames = append(currentlyUnneededNodeNames, node.Name)
+	}
+
+	if skipped > 0 {
+		klog.V(1).Infof("Scale-down calculation: ignoring %v nodes unremovable in the last %v", skipped, c.context.AutoscalingOptions.UnremovableNodeRecheckTimeout)
+	}
+	return currentlyUnneededNodeNames, utilizationMap
+}
+
+func (c *Checker) checkNodeUtilization(timestamp time.Time, nodeInfo *schedulerframework.NodeInfo) (simulator.UnremovableReason, *utilization.Info) {
+	node := nodeInfo.Node()
+
+	// Skip nodes marked to be deleted, if they were marked recently.
+	// Old-time marked nodes are again eligible for deletion - something went wrong with them
+	// and they have not been deleted.
+	if actuation.IsNodeBeingDeleted(node, timestamp) {
```

Comment:
This check has nothing to do with node utilization. Maybe move it to FilterOutUnremovable or rename this function?